cmd/juju/application/refresh.go

```diff
@@ -319,18 +322,30 @@ func (c *refreshCommand) Run(ctx *cmd.Context) error {
 		return errors.Trace(err)
 	}
 	charmRefreshClient := c.NewCharmRefreshClient(apiRoot)
-	oldURL, oldOrigin, err := charmRefreshClient.GetCharmURLOrigin(generation, c.ApplicationName)
-	if err != nil {
-		return errors.Trace(err)
-	}
 
 	// There is a timing window where deploy has been called and the charm
 	// is not yet downloaded. Check here to verify the origin has an ID,
-	// otherwise refresh result may be in accurate. We could use the
-	// retry package, but the issue is only seen in automated test due to
-	// speed. Can always use retry if it becomes an issue.
-	if oldOrigin.Source == commoncharm.OriginCharmHub && oldOrigin.ID == "" {
-		return errors.Errorf("%q deploy incomplete, please try refresh again in a little bit.", c.ApplicationName)
+	// otherwise refresh result may be inaccurate. If there is no ID, we will
+	// retry a couple of times while we wait for the charm to download.
+	var oldURL *charm.URL
+	var oldOrigin commoncharm.Origin
+	err = retry.Call(retry.CallArgs{
+		Func: func() error {
+			oldURL, oldOrigin, err = charmRefreshClient.GetCharmURLOrigin(generation, c.ApplicationName)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if oldOrigin.Source == commoncharm.OriginCharmHub && oldOrigin.ID == "" {
+				return errors.Errorf("failed to get charm ID for %q", c.ApplicationName)
```

Comment:
We should probably keep the original error message here

Reply:
had to change it a little bit - see test changes